packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Juno contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* eslint-disable react/prop-types */
+
+import React, { createContext, useContext, useEffect, useRef, useState } from "react"
+import { Menu as HeadlessMenu } from "@headlessui/react"
+import { Float } from "@headlessui-float/react"
+import { Icon, KnownIconsEnum } from "../Icon/Icon.component"
+import { PortalProvider } from "../PortalProvider/"
+
+// ----- Styles -----
+
+const defaultToggleStyles = `
+  hover:jn-text-theme-accent
+  active:jn-text-theme-accent
+`
+
+const disabledToggleStyles = `
+  jn-cursor-not-allowed
+  jn-opacity-50
+`
+
+const menuStyles = `
+  jn-overflow-hidden
+  jn-flex
+  jn-flex-col
+  jn-w-max
+  jn-rounded
+  jn-bg-theme-background-lvl-1
+`
+
+const itemStyles = `
+  jn-text-base
+  jn-text-theme-default
+  jn-flex
+  jn-w-full
+  jn-items-center
+  jn-whitespace-nowrap
+  jn-pt-[0.6875rem]
+  jn-pb-[0.5rem]
+  jn-px-[0.875rem]
+`
+
+const normalItemStyles = `
+  jn-text-base
+  jn-pt-[0.6875rem]
+  jn-pb-[0.5rem]
+  jn-px-[0.875rem]
+`
+
+const smallItemStyles = `
+  jn-text-sm
+  jn-p-2
+`
+
+const actionableItemStyles = `
+  hover:jn-bg-theme-background-lvl-3
+  jn-cursor-pointer
+`
+
+const disabledItemStyles = `
+  jn-cursor-not-allowed
+  jn-opacity-50
+  jn-bg-theme-background-lvl-1
+`
+
+const itemIconStyles = `
+  jn-inline-block
+  jn-mr-2
+`
+
+const sectionStyles = `
+  jn-border-b
+  jn-border-theme-background-lvl-3
+  last:jn-border-b-0
+`
+
+const sectionTitleStyles = `
+  jn-text-theme-light
+  jn-text-xs
+  jn-p-2
+  jn-cursor-default
+`
+
+const sectionSeparatorStyles = `
+  jn-h-
+  jn-bg-theme-background-lvl-3
+`
+
+// Make sure the floating reference wrapper element around the toggle fits snug around the toggle, so the positioning of the menu is correct relative to the visible toggle:
+const floatingReferenceWrapperStyles = `
+  jn-inline-flex
+`
+
+// ----- Interfaces -----
+
+// Type the PopupMenu context
+export interface PopupMenuContextType {
+  close: () => void
+  isOpen: boolean
+  menuSize: "normal" | "small"
+}
+
+export interface PopupMenuProps extends React.ComponentProps<typeof HeadlessMenu> {
```

Comment:
(question)

Does the HeadlessMenu not providing any types? that's why you use typeof here?

Reply:
My understanding is this is the correct way to use the types provided by Headless Menu and extend them with our own? I am still new to TypeScript and this is by far the most complex component I have ever written in TS, I may be wrong here and happy to learn how if there is a better or more idiomatic way to do this.